src/utils/response.ts

```diff
@@ -86,19 +86,39 @@ export function sendStream (event: H3Event, data: any): Promise<void> {
   })
 }
 
-export function writeEarlyHints (event: H3Event, links: string | string[], callback?: () => void) {
+const noop = () => {}
+export function writeEarlyHints (event: H3Event, hints: string | string[] | Record<string, string | string[]>, cb: () => void = noop) {
   if (!event.res.socket && !('writeEarlyHints' in event.res)) {
-    if (callback) {
-      callback()
-    }
+    cb()
     return
   }
 
+  // Normalize if string or string[] is provided
+  if (typeof hints === 'string' || Array.isArray(hints)) {
+    hints = { link: hints }
+  }
+
   if ('writeEarlyHints' in event.res) {
-    // @ts-expect-error native node 18 implementation
-    return event.res.writeEarlyHints(links, callback)
+    return event.res.writeEarlyHints(hints, cb)
+  }
+
+  const headers: [string, string | string[]][] = Object.entries(hints)
+  if (!headers.length) {
+    cb()
+    return
+  }
+
+  let hint = 'HTTP/1.1 103 Early Hints'
+  if (hints.link) {
+    const links = Array.isArray(hints.link) ? hints.link : Array.from(hints.link)
+    hint += `\r\nLink: ${links.join('\r\n')
+      // TODO: remove when https://github.com/nodejs/node/pull/44874 is released
+      .replace(/; crossorigin/g, '').split(', ')}`
```

Comment:
Shouldn't we join it back before stringification?

Reply:
no. only string is supported for these headers

Reply:
oh, yes. we shouldn't split it here

Reply:
split() retuns an array... Array[].toString format joins string by `,` itself but why we needed split first?

Reply:
we shouldn' tsplit it here at all - it needs to be earlier. see last commit